doc/rst/technotes/initTypeAlias.rst

```diff
@@ -0,0 +1,161 @@
+
+.. _readme-initTypeAlias:
+
+.. default-domain:: chpl
+
+==================================================
+Invoking Initializers with a Generic Instantiation
+==================================================
+
+Overview
+--------
+
+Initializing a variable with a generic type has historically required users to
+pass each ``type`` and ``param`` value to the initializer as arguments. This
+can be a tedious and repetitive process if the user is initializing many
+variables of the exact same type. Some users have expressed interest in support
+for using type aliases to reduce the repetitive nature of initializing the same
+type many times. For example:
+
+.. code-block:: chapel
+
+  record Map { type K, V; ... }
+
+  var x = new Map(string, string, ...);
+
+  // vs.
+
+  type StringMap = Map(string, string);
+  var y = new StringMap(...);
+
+This document describes an experimental solution to this problem that unifies
+the invocation of default-initialization and these kinds of new-expressions.
+
+An Experimental Solution: Using Named Expressions
+-------------------------------------------------
+
+A solution to this problem is for the compiler to generate named-expression
+arguments for each instantiated field in the type alias. This means that if the
+author of a type wishes to support this pattern, they need to implement an
+initializer with arguments that have the same name as the generic fields of the
+type. For example:
+
+.. code-block:: chapel
+
+  record R {
+    type T;
+    var x : T;
+  }
+
+  // must be named 'T'
+  proc R.init(type T, x : T) { ... }
+
+  type RI = R(int);
+  var x = new RI(5);
+  // becomes...
+  //   var x = new R(T=int, 5);
+
+The use of named-expressions is a trade-off between flexibility of argument
+names and flexibility of argument ordering. Consider the following record with
+``type`` fields that are declared adjacent to the fields they constrain:
+
+.. code-block:: chapel
+
+  record R {
+    type T;
+    var x : T;
+    type U;
+    var y : U;
+  }
+
+  // mimics compiler-generated signature
+  // proc R.init(type T, x : T, type U, y : U) { ... }
+
+Without named-expressions, the instantiation information would need to be
+passed as the first arguments of the initializer. This would prevent usage
+of the compiler-generated initializer for types like ``R``:
+
+.. code-block:: chapel
+
+  type RIS = R(int, string);
+  var x = new RIS(5, "hello");
+  // without named-expressions, compiler  cannot match
+  // (type, type, value, value) to (type, value, type, value)
+  //   var x = new R(int, string, 5, "hello");
+
+By prioritizing flexibility of argument ordering, type authors may order their
+fields as they wish and can still take advantage of the compiler-generated
+initializer when using type aliases:
+
+.. code-block:: chapel
+
+  type RIS = R(int, string);
+  var y = new RIS(5, "hello");
+  // becomes...
+  //   var y = new R(T=int, U=string, 5, "hello");
+
+Unification with Default-Initialization
```

Comment:
I'd call it something like "Default Initialization for Instantiated Generic Types"